In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

Reading the diagnosis file

In [2]:
dx = pd.read_csv('dx.csv')

In [3]:
dx.describe()

F_CLAIM_HEADER_CLAIM_NUMBER D_DIAGNOSIS         DESCRIPTION  \
count                                  24632       23809               23809   
unique                                  7589        1763                1779   
top     022f4489-1cdd-50dc-a7b2-05914a609fe4         G35  Multiple sclerosis   
freq                                      30        1523                1523   

                                             D_PATIENT_ID DIAGNOSIS_SEQUENCE  
count                                               24632              24632  
unique                                                300                 27  
top     DeBYjY7+AMJF3SuvyPTXQXwL2rtwvhJHjjww/cA7YZU=Jh...                  1  
freq                                                  801               7602

In [4]:
for col in dx:
    print(dx[col].isnull().sum())

0
823
823
0
0


In [5]:
#Finding the claims for multiple sclerosis
(dx['D_DIAGNOSIS'].str.startswith('G35')).value_counts()

False    22286
True      1523
Name: D_DIAGNOSIS, dtype: int64

In [6]:
#Finding the claims for Breast Cancer
(dx['D_DIAGNOSIS'].str.startswith('C50')).value_counts()

False    21954
True      1855
Name: D_DIAGNOSIS, dtype: int64

In [7]:
#Finding the claims for Wet Amd
(dx['D_DIAGNOSIS'].str.startswith('H3532')).value_counts()

False    23146
True       663
Name: D_DIAGNOSIS, dtype: int64

Number of unique traits for patients diagnosed with wetamd

In [8]:
dx.loc[dx['D_DIAGNOSIS'].str.startswith('H3532',na=False)].nunique()

F_CLAIM_HEADER_CLAIM_NUMBER    511
D_DIAGNOSIS                     12
DESCRIPTION                     12
D_PATIENT_ID                    80
DIAGNOSIS_SEQUENCE              12
dtype: int64

Number of unique traits for patients diagnosed with Breast Cancer

In [9]:
dx.loc[dx['D_DIAGNOSIS'].str.startswith('C50',na=False)].nunique()

F_CLAIM_HEADER_CLAIM_NUMBER    1731
D_DIAGNOSIS                      21
DESCRIPTION                      21
D_PATIENT_ID                     83
DIAGNOSIS_SEQUENCE               12
dtype: int64

Number of unique traits for patients diagnosed with Multiple Sclerosis

In [10]:
dx.loc[dx['D_DIAGNOSIS'].str.startswith('G35',na=False)].nunique()

F_CLAIM_HEADER_CLAIM_NUMBER    1462
D_DIAGNOSIS                       1
DESCRIPTION                       1
D_PATIENT_ID                     81
DIAGNOSIS_SEQUENCE               15
dtype: int64

In [11]:
dx.loc[dx['D_DIAGNOSIS'].isnull()].nunique()

F_CLAIM_HEADER_CLAIM_NUMBER    770
D_DIAGNOSIS                      0
DESCRIPTION                      0
D_PATIENT_ID                    60
DIAGNOSIS_SEQUENCE              11
dtype: int64

# Few Findings

*   The number of patients diagnosed with Wetamd, Breast Cancer & 
    Multiple Sclerosis are 80, 83 & 81 respectively.
*   Multiple Sclerosis has only one unique code and description while wet amd and breast cancer have 12 and 21 unique codes and descriptions respectively.
*   There are 60 unique patients with no diagnosis and description.
*   Since there are only 300 unique patient IDs and adding patients diagnosed with the three diseases and the patients with no diagnosis, the total comes out to be 304, which means some patients have more than one disease.








Reading the header file

In [12]:
hd = pd.read_csv('hd.csv')

In [13]:
hd.describe()

CLAIM_NUMBER  \
count                                   7833   
unique                                  7833   
top     91f66a61-4c68-5426-9437-e05932b0e289   
freq                                       1   

                                             D_PATIENT_ID DATE_OF_SERVICE  
count                                                7833            7833  
unique                                                300             188  
top     GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...      2020-09-01  
freq                                                  378             150

Reading the physician description file

In [14]:
pr = pd.read_csv('pr.csv')

In [15]:
pr.describe()

F_CLAIM_HEADER_CLAIM_NUMBER   HEALTHGRAPH_ID  \
count                                   9949             9949   
unique                                  5940             1857   
top     e90db983-137b-5d4b-ba78-7f941a82de92  P-LND4KKJR9Z0ED   
freq                                       7              159   

                                             D_PATIENT_ID     CLASSIFICATION  \
count                                                9949               9949   
unique                                                291                 58   
top     sHM2QJEJKwQWMM7gLpPg9VVysbiSSoeKoszqT2wjpPA=Yf...  Internal Medicine   
freq                                                  238               3234   

               SPECIALIZATION  
count                    5402  
unique                    102  
top     Hematology & Oncology  
freq                      719

In [17]:
pr.SPECIALIZATION.isna().sum()

4547

Reading the patient description file

In [18]:
pt = pd.read_csv('pt.csv')

In [19]:
pt.describe()

PATIENT_GENDER_CODE PATIENT_DATE_OF_BIRTH  \
count                  299                   293   
unique                   4                    64   
top                      F            1930-01-01   
freq                   190                    20   

                                             D_PATIENT_ID  
count                                                 300  
unique                                                300  
top     ztvWZ+cFYqdyAMVNv/1+LXqktQ7iXuF3Iujwjk+ijPI=ZU...  
freq                                                    1

In [20]:
pt.PATIENT_DATE_OF_BIRTH.isna().sum()

7

Reading Patient treatment description file

In [21]:
px = pd.read_csv('px.csv')

In [22]:
px.describe()

D_PRODUCT_NDC  PROCEDURE_SEQUENCE  UNITS_BILLED    LINE_CHARGE
count   1.694000e+03        20135.000000  20018.000000   19813.000000
mean    4.316028e+10            6.243258     10.567694     884.944380
std     2.686882e+10           11.355168     67.769213    3696.264176
min     3.029328e+06            1.000000      0.000000       0.000000
25%     5.024206e+09            1.000000      1.000000      68.000000
50%     5.551302e+10            2.000000      1.000000     180.000000
75%     6.332305e+10            6.000000      2.000000     421.000000
max     7.632934e+10          142.000000   5560.000000  117015.000000

Merging the diagnosis dataframe with physician description file based on claim numbers

In [23]:
dfn = dx.merge(pr,how='left', left_on='F_CLAIM_HEADER_CLAIM_NUMBER', right_on='F_CLAIM_HEADER_CLAIM_NUMBER',suffixes=('', '_drop'))
dfn.drop([col for col in dfn.columns if 'drop' in col], axis=1, inplace=True)
dfn

F_CLAIM_HEADER_CLAIM_NUMBER D_DIAGNOSIS  \
0      1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
1      e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
2      90a0510e-3117-5d7f-8ad3-63c661894e28         I10   
3      90a0510e-3117-5d7f-8ad3-63c661894e28         I10   
4      03ca39b6-a0a3-5571-a562-a473c86b691c      H43313   
...                                     ...         ...   
37574  9db0449f-b47e-5c7c-bfe4-c1471fb26af1       Z8719   
37575  accf9bd4-e6b3-5de2-b91c-5488dface607       Z8719   
37576  aeafaf7e-bad6-5a91-9e2e-ac7ad1e28071       Z8719   
37577  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
37578  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   

                                             DESCRIPTION  \
0      Age-related osteoporosis without current patho...   
1                                          Other amnesia   
2                       Essential (primary) hypertension   
3                       Essential (primary) hypertension   
4              Vitreous membranes and strands, bilateral   
...                                                  ...   
37574  Personal history of other diseases of the dige...   
37575  Personal history of other diseases of the dige...   
37576  Personal history of other diseases of the dige...   
37577                                Anemia, unspecified   
37578              Other specified postprocedural states   

                                            D_PATIENT_ID DIAGNOSIS_SEQUENCE  \
0      jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
1      uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
2      2mmDd9qcNj5MGEXf0tE8Pu27NBoEpnYUyYn3uj3a51E=R8...                  2   
3      2mmDd9qcNj5MGEXf0tE8Pu27NBoEpnYUyYn3uj3a51E=R8...                  2   
4      X0rr2PkNrCOra2j5TNxPYek/PrSRkdvhg59EzTBJDZM=uf...                  7   
...                                                  ...                ...   
37574  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  9   
37575  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  9   
37576  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  9   
37577  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
37578  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   

        HEALTHGRAPH_ID                                     CLASSIFICATION  \
0       P-QM6N92KMGDPM                                  Internal Medicine   
1      P-078GG9K21P6DE                                    Family Medicine   
2       P-2XP2PX87DMOZ  Student in an Organized Health Care Education/...   
3       P-2XP2PX87DMOZ                                 Emergency Medicine   
4       P-2EL0QZJEK1RK                                      Ophthalmology   
...                ...                                                ...   
37574   P-JOGOO8J07R0K                                          Radiology   
37575   P-JOGOO8J07R0K                                          Radiology   
37576   P-JOGOO8J07R0K                                          Radiology   
37577              NaN                                                NaN   
37578   P-JOGOO8J07R0K                                          Radiology   

           SPECIALIZATION  
0                     NaN  
1                     NaN  
2                     NaN  
3                     NaN  
4                     NaN  
...                   ...  
37574  Radiation Oncology  
37575  Radiation Oncology  
37576  Radiation Oncology  
37577                 NaN  
37578  Radiation Oncology  

[37579 rows x 8 columns]

Merging the previous dataframe with patient treatment description file based on claim numbers


In [24]:
dfn = dfn.merge(px,how='left', left_on='F_CLAIM_HEADER_CLAIM_NUMBER', right_on='F_CLAIM_HEADER_CLAIM_NUMBER',suffixes=('', '_drop'))
dfn.drop([col for col in dfn.columns if 'drop' in col], axis=1, inplace=True)
dfn

F_CLAIM_HEADER_CLAIM_NUMBER D_DIAGNOSIS  \
0       1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
1       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
2       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
3       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
4       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
...                                      ...         ...   
141875  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141876  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141877  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141878  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141879  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   

                                              DESCRIPTION  \
0       Age-related osteoporosis without current patho...   
1                                           Other amnesia   
2                                           Other amnesia   
3                                           Other amnesia   
4                                           Other amnesia   
...                                                   ...   
141875                                Anemia, unspecified   
141876                                Anemia, unspecified   
141877                                Anemia, unspecified   
141878                                Anemia, unspecified   
141879              Other specified postprocedural states   

                                             D_PATIENT_ID DIAGNOSIS_SEQUENCE  \
0       jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
1       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
2       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
3       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
4       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
...                                                   ...                ...   
141875  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141876  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141877  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141878  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141879  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   

         HEALTHGRAPH_ID     CLASSIFICATION      SPECIALIZATION  D_PRODUCT_NDC  \
0        P-QM6N92KMGDPM  Internal Medicine                 NaN            NaN   
1       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
2       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
3       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
4       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
...                 ...                ...                 ...            ...   
141875              NaN                NaN                 NaN            NaN   
141876              NaN                NaN                 NaN            NaN   
141877              NaN                NaN                 NaN            NaN   
141878              NaN                NaN                 NaN            NaN   
141879   P-JOGOO8J07R0K          Radiology  Radiation Oncology            NaN   

       D_PROCEDURE_CODE  PROCEDURE_SEQUENCE  UNITS_BILLED  LINE_CHARGE  
0                   NaN                 NaN           NaN          NaN  
1                 3075F                 6.0           1.0          NaN  
2                 3079F                 5.0           1.0          NaN  
3                 1125F                 2.0           1.0          NaN  
4                 1160F                 4.0           1.0          NaN  
...                 ...                 ...           ...          ...  
141875            83540                 3.0           1.0         19.0  
141876            84466                 8.0  

In [25]:
dfn = dfn[dfn['D_PATIENT_ID'].notna()]
dfn

F_CLAIM_HEADER_CLAIM_NUMBER D_DIAGNOSIS  \
0       1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
1       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
2       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
3       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
4       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
...                                      ...         ...   
141875  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141876  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141877  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141878  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141879  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   

                                              DESCRIPTION  \
0       Age-related osteoporosis without current patho...   
1                                           Other amnesia   
2                                           Other amnesia   
3                                           Other amnesia   
4                                           Other amnesia   
...                                                   ...   
141875                                Anemia, unspecified   
141876                                Anemia, unspecified   
141877                                Anemia, unspecified   
141878                                Anemia, unspecified   
141879              Other specified postprocedural states   

                                             D_PATIENT_ID DIAGNOSIS_SEQUENCE  \
0       jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
1       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
2       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
3       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
4       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
...                                                   ...                ...   
141875  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141876  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141877  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141878  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141879  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   

         HEALTHGRAPH_ID     CLASSIFICATION      SPECIALIZATION  D_PRODUCT_NDC  \
0        P-QM6N92KMGDPM  Internal Medicine                 NaN            NaN   
1       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
2       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
3       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
4       P-078GG9K21P6DE    Family Medicine                 NaN            NaN   
...                 ...                ...                 ...            ...   
141875              NaN                NaN                 NaN            NaN   
141876              NaN                NaN                 NaN            NaN   
141877              NaN                NaN                 NaN            NaN   
141878              NaN                NaN                 NaN            NaN   
141879   P-JOGOO8J07R0K          Radiology  Radiation Oncology            NaN   

       D_PROCEDURE_CODE  PROCEDURE_SEQUENCE  UNITS_BILLED  LINE_CHARGE  
0                   NaN                 NaN           NaN          NaN  
1                 3075F                 6.0           1.0          NaN  
2                 3079F                 5.0           1.0          NaN  
3                 1125F                 2.0           1.0          NaN  
4                 1160F                 4.0           1.0          NaN  
...                 ...                 ...           ...          ...  
141875            83540                 3.0           1.0         19.0  
141876            84466                 8.0  

In [26]:
dfn.loc[dfn['D_DIAGNOSIS'].isnull()].nunique()

F_CLAIM_HEADER_CLAIM_NUMBER    770
D_DIAGNOSIS                      0
DESCRIPTION                      0
D_PATIENT_ID                    60
DIAGNOSIS_SEQUENCE              11
HEALTHGRAPH_ID                 160
CLASSIFICATION                  24
SPECIALIZATION                  32
D_PRODUCT_NDC                  112
D_PROCEDURE_CODE               371
PROCEDURE_SEQUENCE              39
UNITS_BILLED                    63
LINE_CHARGE                    650
dtype: int64

In [27]:
dfn.SPECIALIZATION.isna().mean()

0.5756836763462081

In [28]:
dfn.D_PRODUCT_NDC.isna().mean()

0.8967648717225825

In [29]:
dfn.CLASSIFICATION.isna().mean()

0.09363546659148576

In [30]:
dfn.HEALTHGRAPH_ID.isna().mean()

0.09363546659148576

Since, more than 57 percent entries in Specialization column and 89 percent in the D_Product_NDC column are nan, these attributes cannot be used in predictive modeling and hence will be dropped.



In [31]:
dfn = dfn.drop('SPECIALIZATION', 1)
dfn = dfn.drop('D_PRODUCT_NDC', 1)
dfn

F_CLAIM_HEADER_CLAIM_NUMBER D_DIAGNOSIS  \
0       1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
1       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
2       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
3       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
4       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
...                                      ...         ...   
141875  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141876  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141877  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141878  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141879  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   

                                              DESCRIPTION  \
0       Age-related osteoporosis without current patho...   
1                                           Other amnesia   
2                                           Other amnesia   
3                                           Other amnesia   
4                                           Other amnesia   
...                                                   ...   
141875                                Anemia, unspecified   
141876                                Anemia, unspecified   
141877                                Anemia, unspecified   
141878                                Anemia, unspecified   
141879              Other specified postprocedural states   

                                             D_PATIENT_ID DIAGNOSIS_SEQUENCE  \
0       jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
1       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
2       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
3       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
4       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
...                                                   ...                ...   
141875  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141876  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141877  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141878  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141879  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   

         HEALTHGRAPH_ID     CLASSIFICATION D_PROCEDURE_CODE  \
0        P-QM6N92KMGDPM  Internal Medicine              NaN   
1       P-078GG9K21P6DE    Family Medicine            3075F   
2       P-078GG9K21P6DE    Family Medicine            3079F   
3       P-078GG9K21P6DE    Family Medicine            1125F   
4       P-078GG9K21P6DE    Family Medicine            1160F   
...                 ...                ...              ...   
141875              NaN                NaN            83540   
141876              NaN                NaN            84466   
141877              NaN                NaN            83010   
141878              NaN                NaN            85025   
141879   P-JOGOO8J07R0K          Radiology            77412   

        PROCEDURE_SEQUENCE  UNITS_BILLED  LINE_CHARGE  
0                      NaN           NaN          NaN  
1                      6.0           1.0          NaN  
2                      5.0           1.0          NaN  
3                      2.0           1.0          NaN  
4                      4.0           1.0          NaN  
...                    ...           ...          ...  
141875                 3.0           1.0         19.0  
141876                 8.0           1.0         37.0  
141877                 5.0           1.0         36.0  
141878                 2.0           1.0         23.0  
141879                 1.0           1.0       1611.0  

[141880 rows x 11 columns]

Merging patient description dataframe with the main dataframe

In [32]:
dfn = dfn.merge(pt,how='left', left_on='D_PATIENT_ID', right_on='D_PATIENT_ID',suffixes=('', '_drop'))
dfn.drop([col for col in dfn.columns if 'drop' in col], axis=1, inplace=True)
dfn

F_CLAIM_HEADER_CLAIM_NUMBER D_DIAGNOSIS  \
0       1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
1       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
2       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
3       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
4       e64eecff-c143-52a8-91eb-c04edfdcf226        R413   
...                                      ...         ...   
141875  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141876  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141877  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141878  07a9ca56-0aec-5584-87ae-7911fcba4fda        D649   
141879  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   

                                              DESCRIPTION  \
0       Age-related osteoporosis without current patho...   
1                                           Other amnesia   
2                                           Other amnesia   
3                                           Other amnesia   
4                                           Other amnesia   
...                                                   ...   
141875                                Anemia, unspecified   
141876                                Anemia, unspecified   
141877                                Anemia, unspecified   
141878                                Anemia, unspecified   
141879              Other specified postprocedural states   

                                             D_PATIENT_ID DIAGNOSIS_SEQUENCE  \
0       jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
1       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
2       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
3       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
4       uLIsjMhl634+SIq5r1DNMoyM7DqlPRx2en1zScXyVj4=9A...                  5   
...                                                   ...                ...   
141875  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141876  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141877  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141878  YcFxi7QzLMOQbYrMVBN8vBRQP6igrAKQgi/DY5sovQk=uq...                  1   
141879  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   

         HEALTHGRAPH_ID     CLASSIFICATION D_PROCEDURE_CODE  \
0        P-QM6N92KMGDPM  Internal Medicine              NaN   
1       P-078GG9K21P6DE    Family Medicine            3075F   
2       P-078GG9K21P6DE    Family Medicine            3079F   
3       P-078GG9K21P6DE    Family Medicine            1125F   
4       P-078GG9K21P6DE    Family Medicine            1160F   
...                 ...                ...              ...   
141875              NaN                NaN            83540   
141876              NaN                NaN            84466   
141877              NaN                NaN            83010   
141878              NaN                NaN            85025   
141879   P-JOGOO8J07R0K          Radiology            77412   

        PROCEDURE_SEQUENCE  UNITS_BILLED  LINE_CHARGE PATIENT_GENDER_CODE  \
0                      NaN           NaN          NaN                   F   
1                      6.0           1.0          NaN                   F   
2                      5.0           1.0          NaN                   F   
3                      2.0           1.0          NaN                   F   
4                      4.0           1.0          NaN                   F   
...                    ...           ...          ...                 ...   
141875                 3.0           1.0         19.0                   M   
141876                 8.0           1.0         37.0                   M   
141877                 5.0           1.0         36.0                   M   
141878                 2.0           1.0         23.0                   M  

In [33]:
dfn.nunique()

F_CLAIM_HEADER_CLAIM_NUMBER    7589
D_DIAGNOSIS                    1763
DESCRIPTION                    1779
D_PATIENT_ID                    300
DIAGNOSIS_SEQUENCE               27
HEALTHGRAPH_ID                 1819
CLASSIFICATION                   58
D_PROCEDURE_CODE               1358
PROCEDURE_SEQUENCE              138
UNITS_BILLED                    145
LINE_CHARGE                    2970
PATIENT_GENDER_CODE               4
PATIENT_DATE_OF_BIRTH            64
dtype: int64

In [34]:
dfn.loc[dfn['D_DIAGNOSIS'].isnull()].nunique()

F_CLAIM_HEADER_CLAIM_NUMBER    770
D_DIAGNOSIS                      0
DESCRIPTION                      0
D_PATIENT_ID                    60
DIAGNOSIS_SEQUENCE              11
HEALTHGRAPH_ID                 160
CLASSIFICATION                  24
D_PROCEDURE_CODE               371
PROCEDURE_SEQUENCE              39
UNITS_BILLED                    63
LINE_CHARGE                    650
PATIENT_GENDER_CODE              4
PATIENT_DATE_OF_BIRTH           35
dtype: int64

In [35]:
dfn.D_PATIENT_ID.value_counts()

lewV0qW28cDqUDh45iGa54qFomw+xJGs3pCz7sLyDjM=uIRJ/y/nTRnEhG/LA2z1cgTF75BsakymoYc7tbvgdYs=    14313
V06jdn2JN99pQzuX0xFXQ+vhDBtcFtZ8cVWCv1tf4zw=CR5A/U2FfoyDorn3GYrXQfisQiJ4M7C7mDuGHCrXnt8=     8376
2F8dNh8+T2ZOcPB9M2Lpqb2H9xSUBi8h+Vqt1lqZdG0=dCnlrlSIVVUYbtknzHggvwAxJPqqhFjunaR4hqt+UlI=     4178
W/GbSSP8ovd9s4Vz5g03QWgQNGyuToKQkfLIj6qiCX8=Rpmk9AD4XzXosOQdfXOQcQ8Zp6EotLKSU87++IuMzlw=     4042
+2A3YDHCKwXR63VTG3drPU65i1dAKfL6E43AzVKp/7Y=7ZizRC1EOEglVvr9K8VlYqU2htcB7xUaObwnKngnKJ8=     3302
                                                                                            ...  
+lqp6QTUaxiCBeoFzO5n/Z55jyw6p6Uy28Bf8VTkVK8=XxHae71ouKiWKrPPsejZswrA6bWq0bnbDl0mfUSsb9M=       10
6bOOpVJ3EV06a5EA2awCXVBxn+eddUtHjdeJMObd5M8=0JIb8EM1XmCvHWHD4FYsm6kIBCnzli6YShNXmESueCI=       10
xrV91/T/VjQ05haj1l/PA4ccx5Xpznsxz4PuzNF2f/M=5VFopkMIhNco19i8cetdoJlnTOPBhO+y8EXtqoL30Hs=        8
hoJeBotsalr1kKKQe00EjSuep19J3ynUizlQYgzKalQ=YMUQ7b0fA7jkfWY9YRH/bEkddHuxYmdvu6UxPpI+Hqs=        5
4oWyNEco4BtuNGeUcrW/

Finally merging header file with the dataframe

In [36]:
dfn = dfn.merge(hd,how='left', left_on='D_PATIENT_ID', right_on='D_PATIENT_ID',suffixes=('', '_drop'))
dfn.drop([col for col in dfn.columns if 'drop' in col], axis=1, inplace=True)
dfn

F_CLAIM_HEADER_CLAIM_NUMBER D_DIAGNOSIS  \
0        1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
1        1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
2        1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
3        1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
4        1747b9f6-fdc6-5e5e-bc98-c2c359963aba        M810   
...                                       ...         ...   
8209203  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   
8209204  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   
8209205  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   
8209206  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   
8209207  11c89aed-f6ee-5f6e-a495-39b4ca37fc24      Z98890   

                                               DESCRIPTION  \
0        Age-related osteoporosis without current patho...   
1        Age-related osteoporosis without current patho...   
2        Age-related osteoporosis without current patho...   
3        Age-related osteoporosis without current patho...   
4        Age-related osteoporosis without current patho...   
...                                                    ...   
8209203              Other specified postprocedural states   
8209204              Other specified postprocedural states   
8209205              Other specified postprocedural states   
8209206              Other specified postprocedural states   
8209207              Other specified postprocedural states   

                                              D_PATIENT_ID DIAGNOSIS_SEQUENCE  \
0        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
1        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
2        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
3        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
4        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
...                                                    ...                ...   
8209203  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209204  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209205  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209206  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209207  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   

         HEALTHGRAPH_ID     CLASSIFICATION D_PROCEDURE_CODE  \
0        P-QM6N92KMGDPM  Internal Medicine              NaN   
1        P-QM6N92KMGDPM  Internal Medicine              NaN   
2        P-QM6N92KMGDPM  Internal Medicine              NaN   
3        P-QM6N92KMGDPM  Internal Medicine              NaN   
4        P-QM6N92KMGDPM  Internal Medicine              NaN   
...                 ...                ...              ...   
8209203  P-JOGOO8J07R0K          Radiology            77412   
8209204  P-JOGOO8J07R0K          Radiology            77412   
8209205  P-JOGOO8J07R0K          Radiology            77412   
8209206  P-JOGOO8J07R0K          Radiology            77412   
8209207  P-JOGOO8J07R0K          Radiology            77412   

         PROCEDURE_SEQUENCE  UNITS_BILLED  LINE_CHARGE PATIENT_GENDER_CODE  \
0                       NaN           NaN          NaN                   F   
1                       NaN           NaN          NaN                   F   
2                       NaN           NaN          NaN                   F   
3                       NaN           NaN          NaN                   F   
4                       NaN           NaN          NaN                   F   
...                     ...           ...          ...                 ...   
8209203                 1.0           1.0       1611.0                   F   
8209204                 1.0           1.0       1611.0                   F   
8209205                 1.0           1.0       1611.0                   F   
8209206                 1.0   

Dropping the unnecessary columns and removing duplicate rows

In [37]:
dfn = dfn.drop('F_CLAIM_HEADER_CLAIM_NUMBER', 1)
dfn = dfn.drop('UNITS_BILLED', 1)
dfn = dfn.drop('LINE_CHARGE', 1)
dfn = dfn.drop('CLAIM_NUMBER', 1)
dfn.drop_duplicates()
dfn

D_DIAGNOSIS                                        DESCRIPTION  \
0              M810  Age-related osteoporosis without current patho...   
1              M810  Age-related osteoporosis without current patho...   
2              M810  Age-related osteoporosis without current patho...   
3              M810  Age-related osteoporosis without current patho...   
4              M810  Age-related osteoporosis without current patho...   
...             ...                                                ...   
8209203      Z98890              Other specified postprocedural states   
8209204      Z98890              Other specified postprocedural states   
8209205      Z98890              Other specified postprocedural states   
8209206      Z98890              Other specified postprocedural states   
8209207      Z98890              Other specified postprocedural states   

                                              D_PATIENT_ID DIAGNOSIS_SEQUENCE  \
0        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
1        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
2        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
3        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
4        jS/ju05Bewe+JH3/uAf9xE1AJmpNy37tl0wS+iIT978=NO...                  2   
...                                                    ...                ...   
8209203  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209204  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209205  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209206  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   
8209207  TGXuTpnAHK1C1Oig/S3LsWWYZsC6H5tDZCEOEvnxDZo=dw...                  4   

         HEALTHGRAPH_ID     CLASSIFICATION D_PROCEDURE_CODE  \
0        P-QM6N92KMGDPM  Internal Medicine              NaN   
1        P-QM6N92KMGDPM  Internal Medicine              NaN   
2        P-QM6N92KMGDPM  Internal Medicine              NaN   
3        P-QM6N92KMGDPM  Internal Medicine              NaN   
4        P-QM6N92KMGDPM  Internal Medicine              NaN   
...                 ...                ...              ...   
8209203  P-JOGOO8J07R0K          Radiology            77412   
8209204  P-JOGOO8J07R0K          Radiology            77412   
8209205  P-JOGOO8J07R0K          Radiology            77412   
8209206  P-JOGOO8J07R0K          Radiology            77412   
8209207  P-JOGOO8J07R0K          Radiology            77412   

         PROCEDURE_SEQUENCE PATIENT_GENDER_CODE PATIENT_DATE_OF_BIRTH  \
0                       NaN                   F            1943-01-01   
1                       NaN                   F            1943-01-01   
2                       NaN                   F            1943-01-01   
3                       NaN                   F            1943-01-01   
4                       NaN                   F            1943-01-01   
...                     ...                 ...                   ...   
8209203                 1.0                   F            1951-01-01   
8209204                 1.0                   F            1951-01-01   
8209205                 1.0                   F            1951-01-01   
8209206                 1.0                   F            1951-01-01   
8209207                 1.0                   F            1951-01-01   

        DATE_OF_SERVICE  
0            2020-09-18  
1            2020-08-19  
2            2020-10-05  
3            2020-08-19  
4            2020-10-18  
...                 ...  
8209203      2020-09-17  
8209204      2020-08-20  
8209205      2020-09-08  
8209206      2020-09-11  
8209207      2020-09-18  

[8209208 rows x 11 columns]

In [38]:
df = dfn[dfn['D_DIAGNOSIS'].str.startswith('G35',na=False) | dfn['D_DIAGNOSIS'].str.startswith('C50',na=False) |
         dfn['D_DIAGNOSIS'].str.startswith('H3532',na=False) | dfn['D_DIAGNOSIS'].isnull()]
df = df.drop('DESCRIPTION', 1)
df

D_DIAGNOSIS                                       D_PATIENT_ID  \
3524         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3525         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3526         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3527         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3528         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
...             ...                                                ...   
8160055         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160056         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160057         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160058         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160059         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   

        DIAGNOSIS_SEQUENCE   HEALTHGRAPH_ID     CLASSIFICATION  \
3524                     1  P-YXR9QND04JR6L  Internal Medicine   
3525                     1  P-YXR9QND04JR6L  Internal Medicine   
3526                     1  P-YXR9QND04JR6L  Internal Medicine   
3527                     1  P-YXR9QND04JR6L  Internal Medicine   
3528                     1  P-YXR9QND04JR6L  Internal Medicine   
...                    ...              ...                ...   
8160055                  2              NaN                NaN   
8160056                  2              NaN                NaN   
8160057                  2              NaN                NaN   
8160058                  2              NaN                NaN   
8160059                  2              NaN                NaN   

        D_PROCEDURE_CODE  PROCEDURE_SEQUENCE PATIENT_GENDER_CODE  \
3524               J9217                 2.0                   F   
3525               J9217                 2.0                   F   
3526               J9217                 2.0                   F   
3527               J9217                 2.0                   F   
3528               J9217                 2.0                   F   
...                  ...                 ...                 ...   
8160055            T1019                 1.0                   2   
8160056            T1019                 1.0                   2   
8160057            T1019                 1.0                   2   
8160058            T1019                 1.0                   2   
8160059            T1019                 1.0                   2   

        PATIENT_DATE_OF_BIRTH DATE_OF_SERVICE  
3524               1972-01-01      2020-10-07  
3525               1972-01-01      2020-06-15  
3526               1972-01-01      2020-05-18  
3527               1972-01-01      2020-10-06  
3528               1972-01-01      2020-06-15  
...                       ...             ...  
8160055            1951-01-01      2020-09-27  
8160056            1951-01-01      2020-07-03  
8160057            1951-01-01      2020-05-02  
8160058            1951-01-01      2020-06-17  
8160059            1951-01-01      2020-08-08  

[1398270 rows x 10 columns]

In [39]:
#Obtaining age of each claimaint in days
df[['DATE_OF_SERVICE','PATIENT_DATE_OF_BIRTH']] = df[['DATE_OF_SERVICE','PATIENT_DATE_OF_BIRTH']].apply(pd.to_datetime) #if conversion required
df['AGE'] = (df['DATE_OF_SERVICE'] - df['PATIENT_DATE_OF_BIRTH']).dt.days
df = df.drop('DATE_OF_SERVICE', 1)
df = df.drop('PATIENT_DATE_OF_BIRTH', 1)
df

D_DIAGNOSIS                                       D_PATIENT_ID  \
3524         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3525         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3526         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3527         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3528         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
...             ...                                                ...   
8160055         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160056         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160057         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160058         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160059         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   

        DIAGNOSIS_SEQUENCE   HEALTHGRAPH_ID     CLASSIFICATION  \
3524                     1  P-YXR9QND04JR6L  Internal Medicine   
3525                     1  P-YXR9QND04JR6L  Internal Medicine   
3526                     1  P-YXR9QND04JR6L  Internal Medicine   
3527                     1  P-YXR9QND04JR6L  Internal Medicine   
3528                     1  P-YXR9QND04JR6L  Internal Medicine   
...                    ...              ...                ...   
8160055                  2              NaN                NaN   
8160056                  2              NaN                NaN   
8160057                  2              NaN                NaN   
8160058                  2              NaN                NaN   
8160059                  2              NaN                NaN   

        D_PROCEDURE_CODE  PROCEDURE_SEQUENCE PATIENT_GENDER_CODE      AGE  
3524               J9217                 2.0                   F  17812.0  
3525               J9217                 2.0                   F  17698.0  
3526               J9217                 2.0                   F  17670.0  
3527               J9217                 2.0                   F  17811.0  
3528               J9217                 2.0                   F  17698.0  
...                  ...                 ...                 ...      ...  
8160055            T1019                 1.0                   2  25472.0  
8160056            T1019                 1.0                   2  25386.0  
8160057            T1019                 1.0                   2  25324.0  
8160058            T1019                 1.0                   2  25370.0  
8160059            T1019                 1.0                   2  25422.0  

[1398270 rows x 9 columns]

In [40]:
for col in df:
    print(len(df[col].unique()))

35
300
18
873
41
844
108
5
3301


In [41]:
for col in df:
    print(df[col].isnull().sum())

214069
0
0
288749
288749
33312
33312
561
3761


In [42]:
#Changing the Gender column into two categories 
df['PATIENT_GENDER_CODE'] = df['PATIENT_GENDER_CODE'].map({'1': 'M', '2': 'F','M':'M','F':'F'})
df

D_DIAGNOSIS                                       D_PATIENT_ID  \
3524         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3525         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3526         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3527         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3528         C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
...             ...                                                ...   
8160055         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160056         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160057         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160058         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
8160059         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   

        DIAGNOSIS_SEQUENCE   HEALTHGRAPH_ID     CLASSIFICATION  \
3524                     1  P-YXR9QND04JR6L  Internal Medicine   
3525                     1  P-YXR9QND04JR6L  Internal Medicine   
3526                     1  P-YXR9QND04JR6L  Internal Medicine   
3527                     1  P-YXR9QND04JR6L  Internal Medicine   
3528                     1  P-YXR9QND04JR6L  Internal Medicine   
...                    ...              ...                ...   
8160055                  2              NaN                NaN   
8160056                  2              NaN                NaN   
8160057                  2              NaN                NaN   
8160058                  2              NaN                NaN   
8160059                  2              NaN                NaN   

        D_PROCEDURE_CODE  PROCEDURE_SEQUENCE PATIENT_GENDER_CODE      AGE  
3524               J9217                 2.0                   F  17812.0  
3525               J9217                 2.0                   F  17698.0  
3526               J9217                 2.0                   F  17670.0  
3527               J9217                 2.0                   F  17811.0  
3528               J9217                 2.0                   F  17698.0  
...                  ...                 ...                 ...      ...  
8160055            T1019                 1.0                   F  25472.0  
8160056            T1019                 1.0                   F  25386.0  
8160057            T1019                 1.0                   F  25324.0  
8160058            T1019                 1.0                   F  25370.0  
8160059            T1019                 1.0                   F  25422.0  

[1398270 rows x 9 columns]

In [43]:
df.reset_index(drop=True, inplace=True)
df

D_DIAGNOSIS                                       D_PATIENT_ID  \
0            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
1            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
2            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
4            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
...             ...                                                ...   
1398265         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398266         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398267         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398268         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398269         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   

        DIAGNOSIS_SEQUENCE   HEALTHGRAPH_ID     CLASSIFICATION  \
0                        1  P-YXR9QND04JR6L  Internal Medicine   
1                        1  P-YXR9QND04JR6L  Internal Medicine   
2                        1  P-YXR9QND04JR6L  Internal Medicine   
3                        1  P-YXR9QND04JR6L  Internal Medicine   
4                        1  P-YXR9QND04JR6L  Internal Medicine   
...                    ...              ...                ...   
1398265                  2              NaN                NaN   
1398266                  2              NaN                NaN   
1398267                  2              NaN                NaN   
1398268                  2              NaN                NaN   
1398269                  2              NaN                NaN   

        D_PROCEDURE_CODE  PROCEDURE_SEQUENCE PATIENT_GENDER_CODE      AGE  
0                  J9217                 2.0                   F  17812.0  
1                  J9217                 2.0                   F  17698.0  
2                  J9217                 2.0                   F  17670.0  
3                  J9217                 2.0                   F  17811.0  
4                  J9217                 2.0                   F  17698.0  
...                  ...                 ...                 ...      ...  
1398265            T1019                 1.0                   F  25472.0  
1398266            T1019                 1.0                   F  25386.0  
1398267            T1019                 1.0                   F  25324.0  
1398268            T1019                 1.0                   F  25370.0  
1398269            T1019                 1.0                   F  25422.0  

[1398270 rows x 9 columns]

Relacing all the categorical nan values with the most frequent value and replacing the numerical nan values with its mean

In [44]:
df.HEALTHGRAPH_ID.fillna(df.HEALTHGRAPH_ID.value_counts().index[0],inplace=True)
df.CLASSIFICATION.fillna(df.CLASSIFICATION.value_counts().index[0],inplace=True)
df.D_PROCEDURE_CODE.fillna(df.D_PROCEDURE_CODE.value_counts().index[0],inplace=True)
df.PROCEDURE_SEQUENCE.fillna(df.PROCEDURE_SEQUENCE.value_counts().index[0],inplace=True)
df.PATIENT_GENDER_CODE.fillna(df.PATIENT_GENDER_CODE.value_counts().index[0],inplace=True)
df.AGE.fillna(df.AGE.value_counts().mean(),inplace=True)

In [45]:
for col in df:
    print(df[col].isnull().sum())

214069
0
0
0
0
0
0
0
0


Label Encoding each categorical column

In [135]:
from sklearn.preprocessing import LabelEncoder
lb = LabelEncoder()
df.HEALTHGRAPH_ID = lb.fit_transform(df.HEALTHGRAPH_ID)
df.CLASSIFICATION = lb.fit_transform(df.CLASSIFICATION)
df.D_PROCEDURE_CODE = lb.fit_transform(df.D_PROCEDURE_CODE)
df.PATIENT_GENDER_CODE = lb.fit_transform(df.PATIENT_GENDER_CODE)
df.DIAGNOSIS_SEQUENCE = lb.fit_transform(df.DIAGNOSIS_SEQUENCE)
df

D_DIAGNOSIS                                       D_PATIENT_ID  \
0            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
1            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
2            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
4            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
...             ...                                                ...   
1398265         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398266         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398267         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398268         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398269         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   

         DIAGNOSIS_SEQUENCE  HEALTHGRAPH_ID  CLASSIFICATION  D_PROCEDURE_CODE  \
0                         1             837               7               777   
1                         1             837               7               777   
2                         1             837               7               777   
3                         1             837               7               777   
4                         1             837               7               777   
...                     ...             ...             ...               ...   
1398265                   9             605               7               837   
1398266                   9             605               7               837   
1398267                   9             605               7               837   
1398268                   9             605               7               837   
1398269                   9             605               7               837   

         PROCEDURE_SEQUENCE  PATIENT_GENDER_CODE      AGE  
0                       2.0                    0  17812.0  
1                       2.0                    0  17698.0  
2                       2.0                    0  17670.0  
3                       2.0                    0  17811.0  
4                       2.0                    0  17698.0  
...                     ...                  ...      ...  
1398265                 1.0                    0  25472.0  
1398266                 1.0                    0  25386.0  
1398267                 1.0                    0  25324.0  
1398268                 1.0                    0  25370.0  
1398269                 1.0                    0  25422.0  

[1398270 rows x 9 columns]

In [136]:
pre_df = df[df['D_DIAGNOSIS'].isnull()]
pre_df

D_DIAGNOSIS                                       D_PATIENT_ID  \
3221            NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
3222            NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
3223            NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
3224            NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
3225            NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
...             ...                                                ...   
1397783         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
1397784         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
1397785         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
1397786         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
1397787         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   

         DIAGNOSIS_SEQUENCE  HEALTHGRAPH_ID  CLASSIFICATION  D_PROCEDURE_CODE  \
3221                     12             333               3               506   
3222                     12             333               3               506   
3223                     12             333               3               506   
3224                     12             333               3               506   
3225                     12             333               3               506   
...                     ...             ...             ...               ...   
1397783                   1             138               7               810   
1397784                   1             138               7               810   
1397785                   1             138               7               810   
1397786                   1             138               7               810   
1397787                   1             138               7               810   

         PROCEDURE_SEQUENCE  PATIENT_GENDER_CODE      AGE  
3221                    1.0                    0  31298.0  
3222                    1.0                    0  31300.0  
3223                    1.0                    0  31334.0  
3224                    1.0                    0  31306.0  
3225                    1.0                    0  31332.0  
...                     ...                  ...      ...  
1397783                 1.0                    0  25357.0  
1397784                 1.0                    0  25446.0  
1397785                 1.0                    0  25392.0  
1397786                 1.0                    0  25400.0  
1397787                 1.0                    0  25384.0  

[214069 rows x 9 columns]

In [137]:
train_df = df[~df['D_DIAGNOSIS'].isnull()]

In [138]:
for col in train_df:
    print(len(train_df[col].unique()))

34
240
18
720
39
751
107
2
2781


In [139]:
for col in train_df:
    print(train_df[col].isnull().sum())

0
0
0
0
0
0
0
0
0


In [140]:
train_df

D_DIAGNOSIS                                       D_PATIENT_ID  \
0            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
1            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
2            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
3            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
4            C50411  ITmNkEsAjajLaS8efvo2XStY5obXVjOUlFR5wfAoJho=XP...   
...             ...                                                ...   
1398265         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398266         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398267         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398268         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   
1398269         G35  GWtMunXLIRFMqRYNv4yeISts+S3gIwY3mE8vZCmMq+8=DK...   

         DIAGNOSIS_SEQUENCE  HEALTHGRAPH_ID  CLASSIFICATION  D_PROCEDURE_CODE  \
0                         1             837               7               777   
1                         1             837               7               777   
2                         1             837               7               777   
3                         1             837               7               777   
4                         1             837               7               777   
...                     ...             ...             ...               ...   
1398265                   9             605               7               837   
1398266                   9             605               7               837   
1398267                   9             605               7               837   
1398268                   9             605               7               837   
1398269                   9             605               7               837   

         PROCEDURE_SEQUENCE  PATIENT_GENDER_CODE      AGE  
0                       2.0                    0  17812.0  
1                       2.0                    0  17698.0  
2                       2.0                    0  17670.0  
3                       2.0                    0  17811.0  
4                       2.0                    0  17698.0  
...                     ...                  ...      ...  
1398265                 1.0                    0  25472.0  
1398266                 1.0                    0  25386.0  
1398267                 1.0                    0  25324.0  
1398268                 1.0                    0  25370.0  
1398269                 1.0                    0  25422.0  

[1184201 rows x 9 columns]

In [141]:
x_train = train_df.drop('D_DIAGNOSIS',axis=1)
x_train = x_train.drop('D_PATIENT_ID',axis=1)
x_train

DIAGNOSIS_SEQUENCE  HEALTHGRAPH_ID  CLASSIFICATION  D_PROCEDURE_CODE  \
0                         1             837               7               777   
1                         1             837               7               777   
2                         1             837               7               777   
3                         1             837               7               777   
4                         1             837               7               777   
...                     ...             ...             ...               ...   
1398265                   9             605               7               837   
1398266                   9             605               7               837   
1398267                   9             605               7               837   
1398268                   9             605               7               837   
1398269                   9             605               7               837   

         PROCEDURE_SEQUENCE  PATIENT_GENDER_CODE      AGE  
0                       2.0                    0  17812.0  
1                       2.0                    0  17698.0  
2                       2.0                    0  17670.0  
3                       2.0                    0  17811.0  
4                       2.0                    0  17698.0  
...                     ...                  ...      ...  
1398265                 1.0                    0  25472.0  
1398266                 1.0                    0  25386.0  
1398267                 1.0                    0  25324.0  
1398268                 1.0                    0  25370.0  
1398269                 1.0                    0  25422.0  

[1184201 rows x 7 columns]

In [142]:
y_train = train_df[['D_DIAGNOSIS']]
y_train

D_DIAGNOSIS
0            C50411
1            C50411
2            C50411
3            C50411
4            C50411
...             ...
1398265         G35
1398266         G35
1398267         G35
1398268         G35
1398269         G35

[1184201 rows x 1 columns]

Dividing the Classification tasks into three distinct categories

In [143]:
y_train.loc[y_train['D_DIAGNOSIS'].str.startswith('C50'), 'D_DIAGNOSIS'] = 'C50'
y_train.loc[y_train['D_DIAGNOSIS'].str.startswith('H3532'), 'D_DIAGNOSIS'] = 'H3532'
y_train

D_DIAGNOSIS
0               C50
1               C50
2               C50
3               C50
4               C50
...             ...
1398265         G35
1398266         G35
1398267         G35
1398268         G35
1398269         G35

[1184201 rows x 1 columns]

Label Encoding the Categorical data

In [144]:
y_train.D_DIAGNOSIS = lb.fit_transform(y_train.D_DIAGNOSIS)
y_train

D_DIAGNOSIS
0                  0
1                  0
2                  0
3                  0
4                  0
...              ...
1398265            1
1398266            1
1398267            1
1398268            1
1398269            1

[1184201 rows x 1 columns]

'C50' : 0 , 'G35' : 1 , 'H3532' : 2

In [145]:
x_test = pre_df.drop('D_DIAGNOSIS',axis=1)
x_test = x_test.drop('D_PATIENT_ID',axis=1)
x_test.reset_index(drop=True, inplace=True)
x_test

DIAGNOSIS_SEQUENCE  HEALTHGRAPH_ID  CLASSIFICATION  D_PROCEDURE_CODE  \
0                       12             333               3               506   
1                       12             333               3               506   
2                       12             333               3               506   
3                       12             333               3               506   
4                       12             333               3               506   
...                    ...             ...             ...               ...   
214064                   1             138               7               810   
214065                   1             138               7               810   
214066                   1             138               7               810   
214067                   1             138               7               810   
214068                   1             138               7               810   

        PROCEDURE_SEQUENCE  PATIENT_GENDER_CODE      AGE  
0                      1.0                    0  31298.0  
1                      1.0                    0  31300.0  
2                      1.0                    0  31334.0  
3                      1.0                    0  31306.0  
4                      1.0                    0  31332.0  
...                    ...                  ...      ...  
214064                 1.0                    0  25357.0  
214065                 1.0                    0  25446.0  
214066                 1.0                    0  25392.0  
214067                 1.0                    0  25400.0  
214068                 1.0                    0  25384.0  

[214069 rows x 7 columns]

Dividing the known diagnosis data into training testing set of 70 percent to 30 percent for training and testing the predictive model

In [146]:
from sklearn.model_selection import  train_test_split
iv=x_train.columns.tolist()
idep=x_train[iv].values
dep = y_train[y_train.columns.tolist()].values
x_tr, x_te, y_tr, y_te = train_test_split(idep,dep, test_size=0.30, random_state=0)

In [147]:
x_tr.shape ,y_tr.shape, x_te.shape, y_te.shape

((828940, 7), (828940, 1), (355261, 7), (355261, 1))

Now the feature space becomes a 7 dimensional space with labels being a one dimensional space of multi class classification

In [148]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import SGDClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import mean_squared_error, r2_score
from sklearn import metrics

In [149]:
RF_pipe = make_pipeline(StandardScaler(),RandomForestClassifier())
RF_pipe.fit(x_tr, y_tr)
RF_train_y_pred = RF_pipe.predict(x_te)
print(RF_pipe.score(x_te, y_te))
print('RMSLE:', 100*np.sqrt(metrics.mean_squared_log_error(y_te, RF_train_y_pred)))

1.0
RMSLE: 0.0


In [150]:
DT_pipe = make_pipeline(StandardScaler(),DecisionTreeClassifier(random_state=1))
DT_pipe.fit(x_tr, y_tr)
DT_train_y_pred = DT_pipe.predict(x_te)
print(DT_pipe.score(x_te, y_te))
print('RMSLE:', 100*np.sqrt(metrics.mean_squared_log_error(y_te, DT_train_y_pred)))

0.9999943703361753
RMSLE: 0.16446242585731283


In [151]:
SGD_pipe = make_pipeline(StandardScaler(),SGDClassifier())
SGD_pipe.fit(x_tr, y_tr)
SGD_train_y_pred = SGD_pipe.predict(x_te)
print(SGD_pipe.score(x_te, y_te))

0.8160648086899491


Training three different classifiers and testing it RandomForestClassifier is working at maximum accuracy.So to find the Diagnosis RandomForestClassifier can be used.

In [152]:
iv=x_test.columns.tolist()
idep=x_test[iv].values

y_pre = RF_pipe.predict(idep)

In [154]:
len(y_pre)

214069

In [155]:
pre_df.reset_index(drop=True, inplace=True)
pre_df

D_DIAGNOSIS                                       D_PATIENT_ID  \
0              NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
1              NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
2              NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
3              NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
4              NaN  xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5...   
...            ...                                                ...   
214064         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
214065         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
214066         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
214067         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   
214068         NaN  L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O...   

        DIAGNOSIS_SEQUENCE  HEALTHGRAPH_ID  CLASSIFICATION  D_PROCEDURE_CODE  \
0                       12             333               3               506   
1                       12             333               3               506   
2                       12             333               3               506   
3                       12             333               3               506   
4                       12             333               3               506   
...                    ...             ...             ...               ...   
214064                   1             138               7               810   
214065                   1             138               7               810   
214066                   1             138               7               810   
214067                   1             138               7               810   
214068                   1             138               7               810   

        PROCEDURE_SEQUENCE  PATIENT_GENDER_CODE      AGE  
0                      1.0                    0  31298.0  
1                      1.0                    0  31300.0  
2                      1.0                    0  31334.0  
3                      1.0                    0  31306.0  
4                      1.0                    0  31332.0  
...                    ...                  ...      ...  
214064                 1.0                    0  25357.0  
214065                 1.0                    0  25446.0  
214066                 1.0                    0  25392.0  
214067                 1.0                    0  25400.0  
214068                 1.0                    0  25384.0  

[214069 rows x 9 columns]

In [156]:
pt_id = pre_df.D_PATIENT_ID.unique()

In [157]:
len(pt_id),type(pt_id)

(60, numpy.ndarray)

In [158]:
from collections import Counter
ans={}
for i in range(0,len(pt_id)):
    t = pre_df[(pre_df[['D_PATIENT_ID']] == pt_id[i]).all(1)].index.tolist()
    t1 = y_pre[t]
    data = Counter(t1)
    l = data.most_common(1)[0][0]
    if (l==0):
        ans[pt_id[i]] = 'Breast Cancer'
    elif (l==1):
        ans[pt_id[i]] = 'Multiple Sclerosis'
    elif (l==2):
        ans[pt_id[i]] = 'Wet AMD'

In [159]:
ans

{'xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5AjtafCsGkRaiDG3U08RsOrhxInImpuWig0BlvTrz0=': 'Breast Cancer',
 'FPk2LDZTUrntq/hOoY2yqsDHHmaBEfkwDzUSDf3cnQM=COH2UvO0vVU+EqYZsEYFI/RVdCIE8xivKzYfZOtx8Kc=': 'Breast Cancer',
 '2PYbO46PrGx0D0HkHsEj9rl3fn4jU4REmxvga0DzQtw=vwGivthFvddwXZy9SRXu0w0AExMdlb0HJjGlVs9qwAU=': 'Breast Cancer',
 '54iA/kfIIg5aV+m9U0cqn/JG5ylnYvvP3ZeI124IJZg=33+aeEHtMicWfn8hpYbz1VlVrkv/RY2Y9KvLXoYnVBQ=': 'Breast Cancer',
 'C723AR1uCaM3wduG0oOHKF9Ga1Rk0G0yZtxe3lBP13Y=QIl76L3eK0Hl5E+A56LGobvCY+mNyJO2Wxo0x0jRyUA=': 'Breast Cancer',
 '2mmDd9qcNj5MGEXf0tE8Pu27NBoEpnYUyYn3uj3a51E=R8oxV12Ttf4srV9if9YB8ja6pvUklPN5kVrAJz4CqKs=': 'Multiple Sclerosis',
 'eimA7PGL5Pj56rE/6q8Jw9TmKo1sWdLnCbRvf/NLQPo=3rFg44IUWRWOAPkQBqS9o31Uy/v1ZL4oaXZBG9YsY1Q=': 'Wet AMD',
 'ztvWZ+cFYqdyAMVNv/1+LXqktQ7iXuF3Iujwjk+ijPI=ZUYrhslnthP96PUFErfVuxt1d4oDXMn4WcZSe/Ashg8=': 'Wet AMD',
 'L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O7I8kxGRFC8bOjDLVp4JZe163eHCvBj5wc5U3C8qTM=': 'Breast Cancer',
 'h4O31ohZbYZ5iEO

In [160]:
counts = Counter(ans.values())

In [161]:
counts

Counter({'Breast Cancer': 24, 'Multiple Sclerosis': 17, 'Wet AMD': 19})

The RandomForestClassifier Model predicts Breast Cancer diagnosis more than Wet AMD and Multiple Sclerosis. This happens due to class imbalance present in favour of Breast Cancer as more number of claims are present for breast cancer and very few for Wet AMD.
1855 claims for Breast Cancer Diagnosis,
1523 claims for Multiple Sclerosis Diagnosis &
663 claims for Wet AMD diagnosis

Applying Class balance techniques and retraining

In [162]:
from imblearn.over_sampling import SMOTE

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(x_tr, y_tr)



In [163]:
RF_pipe = make_pipeline(StandardScaler(),RandomForestClassifier())
RF_pipe.fit(x_smote, y_smote)
RF_train_y_pred = RF_pipe.predict(x_te)
print(RF_pipe.score(x_te, y_te))
print('RMSLE:', 100*np.sqrt(metrics.mean_squared_log_error(y_te, RF_train_y_pred)))

1.0
RMSLE: 0.0


In [169]:
metrics.f1_score(y_te,RF_train_y_pred,average='micro')

1.0

In [164]:
iv=x_test.columns.tolist()
idep=x_test[iv].values

y_pre = RF_pipe.predict(idep)

In [165]:
from collections import Counter
ans={}
for i in range(0,len(pt_id)):
    t = pre_df[(pre_df[['D_PATIENT_ID']] == pt_id[i]).all(1)].index.tolist()
    t1 = y_pre[t]
    data = Counter(t1)
    l = data.most_common(1)[0][0]
    if (l==0):
        ans[pt_id[i]] = 'Breast Cancer'
    elif (l==1):
        ans[pt_id[i]] = 'Multiple Sclerosis'
    elif (l==2):
        ans[pt_id[i]] = 'Wet AMD'

In [167]:
ans

{'xZC8rqoNTK+oVxJeQELo6QRWepHPWSrII+SKwyN+LZQ=u5AjtafCsGkRaiDG3U08RsOrhxInImpuWig0BlvTrz0=': 'Breast Cancer',
 'FPk2LDZTUrntq/hOoY2yqsDHHmaBEfkwDzUSDf3cnQM=COH2UvO0vVU+EqYZsEYFI/RVdCIE8xivKzYfZOtx8Kc=': 'Breast Cancer',
 '2PYbO46PrGx0D0HkHsEj9rl3fn4jU4REmxvga0DzQtw=vwGivthFvddwXZy9SRXu0w0AExMdlb0HJjGlVs9qwAU=': 'Breast Cancer',
 '54iA/kfIIg5aV+m9U0cqn/JG5ylnYvvP3ZeI124IJZg=33+aeEHtMicWfn8hpYbz1VlVrkv/RY2Y9KvLXoYnVBQ=': 'Breast Cancer',
 'C723AR1uCaM3wduG0oOHKF9Ga1Rk0G0yZtxe3lBP13Y=QIl76L3eK0Hl5E+A56LGobvCY+mNyJO2Wxo0x0jRyUA=': 'Breast Cancer',
 '2mmDd9qcNj5MGEXf0tE8Pu27NBoEpnYUyYn3uj3a51E=R8oxV12Ttf4srV9if9YB8ja6pvUklPN5kVrAJz4CqKs=': 'Wet AMD',
 'eimA7PGL5Pj56rE/6q8Jw9TmKo1sWdLnCbRvf/NLQPo=3rFg44IUWRWOAPkQBqS9o31Uy/v1ZL4oaXZBG9YsY1Q=': 'Wet AMD',
 'ztvWZ+cFYqdyAMVNv/1+LXqktQ7iXuF3Iujwjk+ijPI=ZUYrhslnthP96PUFErfVuxt1d4oDXMn4WcZSe/Ashg8=': 'Wet AMD',
 'L6LhxMnRrHxJzbqeCc6GQCrjNczUQh+B+KLggDw86Ig=7O7I8kxGRFC8bOjDLVp4JZe163eHCvBj5wc5U3C8qTM=': 'Breast Cancer',
 'h4O31ohZbYZ5iEOG1Nhn/I5HOw

In [166]:
Counter(ans.values())

Counter({'Breast Cancer': 21, 'Wet AMD': 19, 'Multiple Sclerosis': 20})

Applying the class balance technique SMOTE the model is nearer towards the correct prediction of 20 each